File: blackjack-oop/blackjack_oop.js
import shuffle from '../../week1/blackjack/support/shuffle.js'
import Ask from 'https://deno.land/x/ask@1.0.6/mod.ts'
import { getDefaultLogger } from '../../week1/blackjack/support/logging.ts'

// import { playerTurn } from '../../week1/blackjack/blackjack.js'
const defaultLogger = await getDefaultLogger()
const LOSE_MESSAGE = 'You lose!'
const WIN_MESSAGE = 'You win!'
const DRAW_MESSAGE = 'Draw!'

export class Card {
  constructor(rank, suit) {
    this.rank = rank
    this.suit = suit.toUpperCase()
  }

  toString() {
    return `${this.rank}${this.suit}`
  }

  get points() {
    let points = 0
    if (this.rank === 'A') {
      points = 11
    } else if (!parseInt(this.rank) || this.rank == 10) {
      points = 10
    } else {
      points = this.rank
    }
    return parseInt(points)
  }
}

export class Hand {
  constructor(cards) {
    if (!cards.every((card) => card instanceof Card)) {
      throw new TypeError('A Hand can only contain Cards')
    }

    this.cards = cards
  }

  get points() {
    let totalPointsFromHand = 0
    if (this.cards.length === 0) {
      return totalPointsFromHand
    } else if (
      this.cards[0].points == 11 &&
      this.cards[1].points == 11 &&
      this.cards.length == 2
    ) {
      totalPointsFromHand = 21
    } else {
      totalPointsFromHand = this.cards.reduce((accumulator, card) => {
        return accumulator + card.points
      }, 0)
      if (this.cards.length > 5 && totalPointsFromHand < 21) {
        totalPointsFromHand = 21
      }
    }
    return totalPointsFromHand
  }

  addCard(card) {
    this.cards.push(card)
  }

  toString(player) {
    return `${player} hand is ${this.cards}\n(${this.points})`
  }
}

export class Deck {
  constructor() {
    this.cards = this.createDeck()
  }

  createDeck() {
    let deck = []
    const suit = ['S', 'D', 'C', 'H']
    const rank = ['A', 2, 3, 4, 5, 6, 7, 8, 9, 10, 'J', 'Q', 'K']
    for (let i = 0; i < suit.length; i++) {
      for (let j = 0; j < rank.length; j++) {
        deck.push(new Card(rank[j], suit[i]))
      }
    }
    return deck
  }

  draw() {
    return this.cards.shift()
  }

  shuffle(seed = Date.now()) {
    return shuffle(this.cards, seed)
  }
}

class Game {
  constructor(logger = defaultLogger) {
    this.logger = logger
    this.gameDeck = new Deck()
    this.gameDeck.shuffle()
    this.playerHand = new Hand([this.gameDeck.draw(), this.gameDeck.draw()])
    this.isPlayerTurn = true

    this.playerTurn()
  }

  hit(hand) {
    let cardDrawn = this.gameDeck.draw()
    hand.addCard(cardDrawn)
    if (this.isPlayerTurn) {
      this.logger.info(`You drew ${cardDrawn}`)
    } else {
      this.logger.info(`Dealer drew ${cardDrawn}`)
    }
  }

  async playerTurn() {
    while (this.isPlayerTurn) {
      this.logger.info(this.playerHand.toString('Your'))
      const ask = new Ask()
      let { action } = await ask.input({
        name: 'action',
        message: 'What do you want to do? ("hit" or "stick")'
      })

      switch (action) {
        case 'hit':
          this.logger.info('hitting')
          this.hit(this.playerHand)
          this.isPlayerTurn = this.checkPlayerLost()
          break
        case 'stick':
          this.isPlayerTurn = false
          this.dealersTurn()
          break
        default:
          continue
      }
    }
    this.checkWhoWon()
  }

  checkPlayerLost() {
    let hasNotLost = true

    if (this.playerHand.points > 21) {
      this.logger.info(this.playerHand.toString('Your'))
      this.logger.info(LOSE_MESSAGE)
      hasNotLost = false
    }

    return hasNotLost
  }

  dealersTurn() {
    this.dealerHand = new Hand([this.gameDeck.draw(), this.gameDeck.draw()])
    this.logger.info(this.dealerHand.toString('Dealers'))
    while (this.dealerHand.points < 17) {
      this.logger.info('hitting')
      this.hit(this.dealerHand)
      this.logger.info(this.dealerHand.toString('Dealers'))
    }
  }

  checkWhoWon() {
    const playerScore = this.playerHand.points
    const dealerScore = this.dealerHand.points

    if (playerScore === dealerScore) {
      this.logger.info(DRAW_MESSAGE)
    } else if (playerScore > dealerScore || dealerScore > 21) {
      this.logger.info(WIN_MESSAGE)
    } else if (dealerScore > playerScore) {
      this.logger.info(LOSE_MESSAGE)
    }
  }
}

// this.dealerHand = new Hand([gameDeck.draw(), gameDeck.draw()])
let blackjack = new Game()

File: blackjack-python/blackjack_oop.py
from functools import reduce 
import random

LOSE_MESSAGE = 'You lose!'
WIN_MESSAGE = 'You win!'
DRAW_MESSAGE = 'Draw!'

class Card:
  def __init__(self, rank, suit):
        self.rank = rank
        self.suit = suit

  def to_string(self):
    return f'{self.rank}{self.suit}'

  @property
  def get_points(self):
    points = 0
    if self.rank == 'A':
      points = 11
    elif type(self.rank) != int or self.rank == 10:
      points = 10
    else:
      points = self.rank
    return points

class Hand:
  def __init__(self, deck):
    self.deck = deck
    self.cards_in_hand = []
    self.draw_card()
    self.draw_card()

  @property
  def get_points(self):
    total_points_in_hand = 0
    if len(self.cards_in_hand) >= 2:
      if self.cards_in_hand[0].get_points == 11 and self.cards_in_hand[1].get_points == 11:
        total_points_in_hand = 21
      else:
        total_points_in_hand = reduce((lambda a, b: a + b.get_points), self.cards_in_hand, 0)
    if len(self.cards_in_hand) > 5 and total_points_in_hand < 21:
      total_points_in_hand = 21
    return total_points_in_hand

  def draw_card(self):
    self.cards_in_hand.append(self.deck.draw())

  def print_cards(self, cards):
    return ', '.join([card.to_string() for card in cards])

  @property
  def latest_card(self):
    return self.cards_in_hand[len(self.cards_in_hand) - 1]

  def to_string(self, player):
    return f'{player} hand is {self.print_cards(self.cards_in_hand)}\n({self.get_points})'

class Deck:
  def __init__(self):
      self.cards = self.create_deck()

  def create_deck(self):
    deck = []
    suits = ['S', 'D', 'C', 'H']
    ranks = ['A', 2, 3, 4, 5, 6, 7, 8, 9, 10, 'J', 'Q', 'K']
    for suit in suits:
      for rank in ranks:
        deck.append(Card(rank, suit))
    return deck

  def get_cards(self):
    return self.cards
  
  def draw(self):
    return self.cards.pop(0)

  def shuffle(self):
    return random.shuffle(self.cards)

class Game:
  def __init__(self):
      self.game_deck = Deck()
      self.game_deck.shuffle()
      self.player_hand = Hand(self.game_deck)
      self.is_player_turn = True
      self.player_turn()

  def player_turn(self):
    while self.is_player_turn == True:
      print(self.player_hand.to_string('Your'))
      print('What do you want to do? ("hit" or "stick")')
      answer = input()
      if answer == 'hit':
        print('Hitting')
        self.hit(self.player_hand)
        self.is_player_turn = self.check_player_lost()
        break
      elif answer == 'stick':
        self.is_player_turn = False
        self.dealers_turn()
        break
      else:
        continue

  def dealers_turn(self):
    self.dealer_hand = Hand(self.game_deck)
    print(self.dealer_hand.to_string('Dealers'))
    while self.dealer_hand.get_points < 17:
      print('Hitting')
      self.hit(self.dealer_hand)
      print(self.dealer_hand.to_string('Dealers'))
    self.check_who_won()
  
  def hit(self, hand):
    hand.draw_card()
    if self.is_player_turn:
      print(f'You drew {hand.latest_card.to_string()}')
    else:
      print(f'Dealer drew {hand.latest_card.to_string()}')

  def check_player_lost(self):
    has_not_lost = True
    if self.player_hand.get_points > 21:
      print(self.player_hand.to_string('Your'))
      print(LOSE_MESSAGE)
      has_not_lost = False
    return has_not_lost
  
  def check_who_won(self):
    player_score = self.player_hand.get_points
    dealer_score = self.dealer_hand.get_points
    if player_score == dealer_score:
      print(DRAW_MESSAGE)
    elif player_score > dealer_score or dealer_score > 21:
      print(WIN_MESSAGE)
    elif dealer_score > player_score:
      print(LOSE_MESSAGE)



blackjack = Game()